### PR TITLE
feat: add linux CSE tracing and metrics  

### LINKED ISSUE #2248: Instrument CSE scripts for Linux
// TODO @devigned

### PR DESCRIPTION
**Pull request by devigned**
**Reason for Change**:

It would help development if we were able to have more insight into the behavior of aks-engine.

This PR introduces a new command line tool for sending telemetry to Application Insights, [apmz](https://github.com/devigned/apmz) (docs in progress). `apmz` outputs each trace and metric to an intermediate file, then flushes them at the end. This behavior is only activated when the feature is turned on in the API model. When the feature is not active, there are still stub no-op bash functions available.

### To query this data, you can run the following kql
```
customMetrics
| where customDimensions.os_type == "linux" and customDimensions.correlation_id == "41b61b03-ae2d-495d-a0f7-869afb024ed3"
| order by timestamp asc 
```

**Issue Fixed**:

Fixes #2248 

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

**Comment by marosset** [comment]
@devigned i opened a new PR with only the apimodel changes for enabling telemetry. Let's try and get that merged and we can work on enabling telemetry for linux and windows CSEs independently
 https://github.com/Azure/aks-engine/pull/2417

**Comment by marosset** [comment]
this PR is much cleaner now :)

File: parts/k8s/cloud-init/artifacts/cse_helpers.sh [hunk 1]
```diff
@@ -55,6 +55,7 @@ ERR_CIS_APPLY_PASSWORD_CONFIG=115 {{/* Error applying CIS-recommended passwd con
 
 ERR_VHD_FILE_NOT_FOUND=124 {{/* VHD log file not found on VM built from VHD distro */}}
 ERR_VHD_BUILD_ERROR=125 {{/* Reserved for VHD CI exit conditions */}}
+ERR_APMZ_INSTALL_ERROR=126 {{/* Unable to install apmz telemetry client */}}
```
Comment (marosset) [inline parts/k8s/cloud-init/artifacts/cse_helpers.sh]: Are you planning on having this be a fatal error?
I've been debating what we should do if we can't get the app insights bins on the Windows side.

Comment (devigned) [reply parts/k8s/cloud-init/artifacts/cse_helpers.sh]: Probably, not.

File: parts/k8s/cloud-init/artifacts/cse_install.sh [hunk 2]
```diff
@@ -177,6 +178,17 @@ downloadContainerd() {
     retrycmd_get_tarball 120 5 "$CONTAINERD_DOWNLOADS_DIR/${CONTAINERD_TGZ_TMP}" ${CONTAINERD_DOWNLOAD_URL} || exit $ERR_CONTAINERD_DOWNLOAD_TIMEOUT
 }
 
+installAPMZ() {
```
Comment (CecileRobertMichon) [inline parts/k8s/cloud-init/artifacts/cse_install.sh]: are we going to pre-install this on the vhd? Also, does uploading metrics require outbound internet? if so, should be disable it/skip it in "locked down environments"?

Comment (devigned) [reply parts/k8s/cloud-init/artifacts/cse_install.sh]: I'd like to pre-install. What do you think about that?

File: pkg/engine/templates_generated.go [hunk 3]
```diff
@@ -13357,6 +13401,21 @@ source {{GetCSEHelpersScriptFilepath}}
...
+function exitAndFlush {
+  if [[ $? != "0" ]]; then
+    trace_err "cse-exit" code=$?
+  else
+    trace_info "cse-exit" code=$?
+  fi
+  apmz batch -f ${TMP_APMZ_BATCH_FILE} --api-key {{GetApplicationInsightsTelemetryKey}}
+}
+
+trap exitAndFlush EXIT
```
Comment (CecileRobertMichon) [inline pkg/engine/templates_generated.go]: this is 💯 

I meant to do something like this a while ago and actually didn't know you could do it like this, just learned something

File: pkg/engine/templates_generated.go [hunk 4]
```diff
@@ -13357,6 +13401,21 @@ source {{GetCSEHelpersScriptFilepath}}
...
+{{- if HasTelemetryEnabled }}
+installAPMZ
+
+function exitAndFlush {
+  if [[ $? != "0" ]]; then
+    trace_err "cse-exit" code=$?
+  else
+    trace_info "cse-exit" code=$?
+  fi
+  apmz batch -f ${TMP_APMZ_BATCH_FILE} --api-key {{GetApplicationInsightsTelemetryKey}}
```
Comment (CecileRobertMichon) [inline pkg/engine/templates_generated.go]: does this ever fail intermittently? Worth retrying?

File: pkg/engine/templates_generated.go [hunk 5]
```diff
@@ -13176,6 +13208,17 @@ downloadContainerd() {
     retrycmd_get_tarball 120 5 "$CONTAINERD_DOWNLOADS_DIR/${CONTAINERD_TGZ_TMP}" ${CONTAINERD_DOWNLOAD_URL} || exit $ERR_CONTAINERD_DOWNLOAD_TIMEOUT
 }
 
+installAPMZ() {
+  apmz_filepath=/usr/local/bin/apmz; apmz_url="https://github.com/devigned/apmz/releases/download/v0.0.1/apmz_v0.0.1_linux_amd64.tar.gz"
```
Comment (CecileRobertMichon) [inline pkg/engine/templates_generated.go]: why is this download path in your fork?

File: parts/k8s/cloud-init/artifacts/cse_helpers.sh (hunk 1 above)
Comment (CecileRobertMichon) [reply parts/k8s/cloud-init/artifacts/cse_helpers.sh]: +1 on not making it fatal, right now looks like it is

File: parts/k8s/cloud-init/artifacts/cse_helpers.sh [hunk 6]
```diff
@@ -55,6 +55,7 @@ ERR_CIS_APPLY_PASSWORD_CONFIG=115 {{/* Error applying CIS-recommended passwd con
 
 ERR_VHD_FILE_NOT_FOUND=124 {{/* VHD log file not found on VM built from VHD distro */}}
 ERR_VHD_BUILD_ERROR=125 {{/* Reserved for VHD CI exit conditions */}}
```
Comment (CecileRobertMichon) [inline parts/k8s/cloud-init/artifacts/cse_helpers.sh]: 120s are reserved for VHD-related errors, let's make the new code 130+

File: parts/k8s/cloud-init/artifacts/cse_install.sh (hunk 2 above)
Comment (CecileRobertMichon) [reply parts/k8s/cloud-init/artifacts/cse_install.sh]: actually we can probably add `{{- if HasTelemetryEnabled }}` here

File: parts/k8s/cloud-init/artifacts/cse_install.sh [hunk 7]
```diff
@@ -177,6 +178,17 @@ downloadContainerd() {
...
+  apmz_filepath=/usr/local/bin/apmz; apmz_url="https://github.com/devigned/apmz/releases/download/v0.0.1/apmz_v0.0.1_linux_amd64.tar.gz"
```
Comment (CecileRobertMichon) [inline parts/k8s/cloud-init/artifacts/cse_install.sh]: why is this download path in your fork?

File: parts/k8s/cloud-init/artifacts/cse_main.sh [hunk 8]
```diff
@@ -19,6 +20,21 @@ source {{GetCSEHelpersScriptFilepath}}
...
+{{- if HasTelemetryEnabled }}
+installAPMZ
+
+function exitAndFlush {
+  if [[ $? != "0" ]]; then
+    trace_err "cse-exit" code=$?
+  else
+    trace_info "cse-exit" code=$?
+  fi
+  apmz batch -f ${TMP_APMZ_BATCH_FILE} --api-key {{GetApplicationInsightsTelemetryKey}}
```
Comment (CecileRobertMichon) [inline parts/k8s/cloud-init/artifacts/cse_main.sh]: does this ever fail intermittently? Worth retrying?

File: parts/k8s/cloud-init/artifacts/cse_main.sh [hunk 9]
```diff
@@ -19,6 +20,21 @@ source {{GetCSEHelpersScriptFilepath}}
...
+}
+
+trap exitAndFlush EXIT
```
Comment (CecileRobertMichon) [inline parts/k8s/cloud-init/artifacts/cse_main.sh]: this is 💯

I meant to do something like this a while ago and actually didn't know you could do it like this, just learned something

File: parts/k8s/cloud-init/artifacts/cse_main.sh (hunk 8 above)
Comment (devigned) [reply parts/k8s/cloud-init/artifacts/cse_main.sh]: retry is handled within apmz

File: parts/k8s/cloud-init/artifacts/cse_install.sh (hunk 7 above)
Comment (devigned) [reply parts/k8s/cloud-init/artifacts/cse_install.sh]: Well... I wrote `apmz`. I'm happy to move it to Azure if we like what's going on with it :).

Comment (CecileRobertMichon) [reply parts/k8s/cloud-init/artifacts/cse_install.sh]: aaah nvm :)

File: parts/k8s/cloud-init/artifacts/cse_install.sh (hunk 2 above)
Comment (marosset) [reply parts/k8s/cloud-init/artifacts/cse_install.sh]: +1 for pre-installing

Comment (marosset) [reply parts/k8s/cloud-init/artifacts/cse_install.sh]: But also keep the install path in there if customers are using other distros

File: parts/k8s/cloud-init/artifacts/cse_helpers.sh [hunk 10]
```diff
@@ -265,4 +265,34 @@ sysctl_reload() {
...
+}
+{{else}}
+trace_err() {
+    return
+}
+trace_info() {
+    return
+}
+time_metric() {
+    return
```
Comment (devigned) [inline parts/k8s/cloud-init/artifacts/cse_helpers.sh]: This should be a call to `"$@"` rather than return.

File: parts/k8s/cloud-init/artifacts/cse_install.sh (hunk 2 above)
Comment (devigned) [reply parts/k8s/cloud-init/artifacts/cse_install.sh]: `apmz` is now pre-installed and will check to verify the version installed matches the version symlinked.

File: examples/kubernetes.json [hunk 11]
```diff
@@ -29,6 +29,9 @@
     "servicePrincipalProfile": {
       "clientId": "",
       "secret": ""
+    },
+    "featureFlags": {
+      "enableTelemetry": true
```
Comment (CecileRobertMichon) [inline examples/kubernetes.json]: I think we want to revert this change to the main example file before the PR is merged

Comment (devigned) [reply examples/kubernetes.json]: I left it there intentionally, but in hindsight, I think you are right. I'll remove it.

**Comment by devigned** [comment]
/hold

Going to move `apmz` into acs-mirror and then change the install path.

**Comment by devigned** [comment]
This PR is only waiting on getting ampz v0.4.0 into the prod storage bucket. Right now, it points to the staging bucket. Once I update it to point at the prod bucket, I'll remove the WIP tag.

@jchauncey, let me know if you need any additional info to approve the prod deploy jobs.

**Comment by devigned** [comment]
@marosset, please give this another look when you have a sec.

File: parts/k8s/cloud-init/artifacts/cse_main.sh [hunk 12]
```diff
@@ -39,22 +47,23 @@ fi
...
 if [ -f /var/run/reboot-required ]; then
     REBOOTREQUIRED=true
+    trace_info "RebootRequired" "reboot=true"
 else
     REBOOTREQUIRED=false
 fi
 
 {{- if not NeedsContainerd}}
-cleanUpContainerd
+time_metric "CleanupContainerd" cleanUpContainerd
```
Comment (jackfrancis) [inline parts/k8s/cloud-init/artifacts/cse_main.sh]: @palma21 FYI, AKS Engine CSE is now instrumented for App Insights telemetry. The `time_metric` func above is a wrapper around all CSE step executions, which is disabled by default, and is a no-op.

Just FYI.
